cdap-api/src/main/java/co/cask/cdap/api/messaging/MessageFetcher.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.messaging;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+
+/**
+ * Provides access to fetch messages from the transactional messaging system.
+ */
+@Beta
+public interface MessageFetcher {
+
+  /**
+   * Fetch messages from the given topic that was published on or after the given timestamp.
+   *
+   * @param namespace namespace of the topic
+   * @param topic name of the topic
+   * @param limit maximum number of messages to fetch
+   * @param timestamp timestamp in milliseconds; the publish timestamp of the first {@link Message}
+   *                  returned by the resulting {@link CloseableIterator} will be on or after the given timestamp.
+   *                  Use {@code timestamp = 0} to poll from the first available message.
+   * @return a {@link CloseableIterator} of {@link Message}
+   * @throws IOException if failed to communicate with the messaging system.
+   * @throws IllegalArgumentException if the give topic doesn't exist or if the timestamp is negative.
+   */
+  CloseableIterator<Message> fetch(String namespace, String topic, int limit, long timestamp) throws IOException;
+
+  /**
+   * Fetch messages from the given topic that was published after a message, identified by the given
+   * message id.
+   *
+   * @param namespace namespace of the topic
+   * @param topic name of the topic
+   * @param limit maximum number of messages to fetch
+   * @param afterMessageId message id returned from the {@link Message#getId()} method via some prior call to
+   *                       one of the {@code pollMessages} methods. If it is {@code null}, it will fetch from
+   *                       the first available message.
```

Comment:
First available message means the message that has the smallest message ID that is currently available in the topic (messages are order by messageId, which has publish timestamp and sequence id in there for uniqueness). However, it is not the first message ever published, because topic messages as TTL. Probably need to add more doc to the `package-info.java` to describe the TMS behavior.